mlx/ops.cpp

```diff
@@ -2222,6 +2222,176 @@ array matmul(
   return out;
 }
 
+/** Computes: O = softmax(Q @ K.T) @ V; with parallelization across K-V sequence
+ * length **/
+array fast_inference_sdpa(
+    const array& queries,
+    const array& keys,
+    const array& values,
+    const float scale,
+    const std::optional<array>& mask,
+    StreamOrDevice s) {
+  auto q = queries;
+  auto k = keys;
+  auto v = values;
+  auto m = mask;
+
+  if (q.ndim() == 0 || k.ndim() == 0 || v.ndim() == 0) {
+    throw std::invalid_argument(
+        "[fast_inference_sdpa] Got 0 dimension input. Inputs must "
+        "have at least one dimension.");
+  }
+  for (const auto& tensor : {q, k, v}) {
+    if (tensor.ndim() != 4) {
+      std::ostringstream msg;
+      msg << "[fast_inference_sdpa] input with shape " << tensor.shape()
+          << " expected to be rank 4";
+      throw std::invalid_argument(msg.str());
+    }
+  }
+
+  size_t batch_dim = q.shape(0);
+  if (batch_dim != 1) {
+    std::ostringstream msg;
+    msg << "[fast_inference_sdpa] Expected batch dimension of 1. ";
+    throw std::invalid_argument(msg.str());
+  }
+  for (const auto& tensor : {q, k, v}) {
+    if (tensor.shape(0) != batch_dim) {
+      std::ostringstream msg;
+      msg << "[fast_inference_sdpa] mismatching batch dimension for input with shape "
+          << tensor.shape() << ".";
+      throw std::invalid_argument(msg.str());
+    }
+  }
+
+  // Below shape checking assumes following tensor shapes:
+  // Q ~ [B H R D], B ~ batch, H ~ n_heads, R ~ query sequence length (1), D ~
+  // d_k (e.g. 128) K ~ [B H_kv L D], B ~ batch, H ~ n_kv_heads, L ~ kv sequence
+  // length (cache size + 1), D ~ d_k V ~ [B H_kv L D_v], B ~ batch, H ~
+  // n_kv_heads, L ~ kv sequence length (cache size + 1), D ~ d_v; Note that for
+  // Multihead attention, n_heads == n_kv_heads; Multi-query attention,
+  // n_kv_heads == 1; and for Grouped Query Attention, n_kv_heads < n_heads,
+  // n_kv_heads > 1, n_heads % n_kv_heads == 0. K, V tensors must have same
+  // number of heads.
+
+  // Q, K must have matching last dims (d_k);
+  if (q.shape(-1) != k.shape(-1)) {
+    std::ostringstream msg;
+    msg << "[fast_inference_sdpa] query, keys expected to have matching d_k; found query "
+        << q.shape() << " for keys " << k.shape() << ".";
```

Comment:
Nit: the `d_k` here is obscure. Maybe replace with something like:

```suggestion
    msg << "[fast_inference_sdpa] query, keys expected to have matching last dimension, but got query shape "
        << q.shape() << " for keys shape " << k.shape() << ".";
```

Reply:
done